<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#exploring-backblaze's-hard-disk-SMART-stats-dataset" data-toc-modified-id="exploring-backblaze's-hard-disk-SMART-stats-dataset-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>exploring backblaze's hard disk SMART stats dataset</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#more-info-from-the-website-regarding-interpretation:" data-toc-modified-id="more-info-from-the-website-regarding-interpretation:-1.0.1"><span class="toc-item-num">1.0.1&nbsp;&nbsp;</span>more info from the website regarding interpretation:</a></span></li></ul></li><li><span><a href="#important-points:" data-toc-modified-id="important-points:-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>important points:</a></span></li></ul></li><li><span><a href="#dropping-null-columns" data-toc-modified-id="dropping-null-columns-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>dropping null columns</a></span><ul class="toc-item"><li><span><a href="#we-should-also-take-care-not-to-drop-any-columns-with-failure=1,-since-we've-got-such-a-class-imbalance-to-deal-with" data-toc-modified-id="we-should-also-take-care-not-to-drop-any-columns-with-failure=1,-since-we've-got-such-a-class-imbalance-to-deal-with-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>we should also take care not to drop any columns with failure=1, since we've got such a class imbalance to deal with</a></span></li><li><span><a href="#98/126-SMART-cols-have->20%-nulls" data-toc-modified-id="98/126-SMART-cols-have->20%-nulls-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>98/126 SMART cols have &gt;20% nulls</a></span></li></ul></li></ul></div>

# exploring backblaze's hard disk SMART stats dataset
https://www.backblaze.com/b2/hard-drive-test-data.html#downloading-the-raw-hard-drive-test-data

In [2]:
import pandas as pd
import numpy as np

import os
from os import listdir
from os.path import isfile, join

import matplotlib.pyplot as pls
import seaborn as sns
%matplotlib inline



i should really memorize how to list all CSVs in a directory using OS...

In [ ]:

onlyfiles = [file for file in listdir(os.getcwd()) if isfile(join(os.getcwd(), file)) and 'csv' in file]

In [5]:
path = os.getcwd()+'/data/'
csvs = [file for file in listdir(path) if isfile(join(path, file)) and 'csv' in file]
# [f for f in listdir(path) if isfile(join(path, file)) and 'csv' in file]
# memorable enough.
csvs

['2020-05-27.csv',
 '2020-05-26.csv',
 '2020-05-18.csv',
 '2020-05-30.csv',
 '2020-05-24.csv',
 '2020-05-25.csv',
 '2020-05-31.csv',
 '2020-05-19.csv',
 '2020-05-21.csv',
 '2020-05-09.csv',
 '2020-05-08.csv',
 '2020-05-20.csv',
 '2020-05-22.csv',
 '2020-05-23.csv',
 '2020-06-15.csv',
 '2020-06-01.csv',
 '2020-06-29.csv',
 '2020-04-10.csv',
 '2020-04-04.csv',
 '2020-04-05.csv',
 '2020-04-11.csv',
 '2020-06-28.csv',
 '2020-06-14.csv',
 '2020-06-02.csv',
 '2020-06-16.csv',
 '2020-04-07.csv',
 '2020-04-13.csv',
 '2020-04-12.csv',
 '2020-04-06.csv',
 '2020-06-17.csv',
 '2020-06-03.csv',
 '2020-06-07.csv',
 '2020-06-13.csv',
 '2020-04-02.csv',
 '2020-04-16.csv',
 '2020-04-17.csv',
 '2020-04-03.csv',
 '2020-06-12.csv',
 '2020-06-06.csv',
 '2020-06-10.csv',
 '2020-06-04.csv',
 '2020-04-29.csv',
 '2020-04-15.csv',
 '2020-04-01.csv',
 '2020-04-14.csv',
 '2020-04-28.csv',
 '2020-06-05.csv',
 '2020-06-11.csv',
 '2020-06-20.csv',
 '2020-06-08.csv',
 '2020-04-25.csv',
 '2020-04-19.csv',
 '2020-04-18

In [ ]:
len(csvs) # 12 million rows isn't even that bad, really...

In [6]:
dpath = path+csvs[0]
df = pd.read_csv(dpath)

In [7]:
df.head()

date   serial_number                 model  capacity_bytes  failure  \
0  2020-05-27        Z305B2QN           ST4000DM000   4000787030016        0   
1  2020-05-27        ZJV0XJQ4         ST12000NM0007  12000138625024        0   
2  2020-05-27        ZJV0XJQ3         ST12000NM0007  12000138625024        0   
3  2020-05-27        ZJV0XJQ0         ST12000NM0007  12000138625024        0   
4  2020-05-27  PL1331LAHG1S4H  HGST HMS5C4040ALE640   4000787030016        0   

   smart_1_normalized  smart_1_raw  smart_2_normalized  smart_2_raw  \
0               117.0  147920672.0                 NaN          NaN   
1                71.0   13356696.0                 NaN          NaN   
2                84.0  241290192.0                 NaN          NaN   
3                83.0  201326680.0                 NaN          NaN   
4               100.0          0.0               134.0        103.0   

   smart_3_normalized  ...  smart_250_normalized  smart_250_raw  \
0                91.0  ...                   NaN            NaN   
1                89.0  ...                   NaN            NaN   
2                98.0  ...                   NaN            NaN   
3                93.0  ...                   NaN            NaN   
4               143.0  ...                   NaN            NaN   

   smart_251_normalized  smart_251_raw  smart_252_normalized  smart_252_raw  \
0                   NaN            NaN                   NaN            NaN   
1                   NaN            NaN                   NaN            NaN   
2                   NaN            NaN                   NaN            NaN   
3                   NaN            NaN                   NaN            NaN   
4                   NaN            NaN                   NaN            NaN   

   smart_254_normalized  smart_254_raw  smart_255_normalized  smart_255_raw  
0                   NaN            NaN                   NaN            NaN  
1                   NaN            NaN                   NaN            NaN  
2                   NaN            NaN                   NaN            NaN  
3                   NaN            NaN                   NaN            NaN  
4                   NaN            NaN                   NaN            NaN  

[5 rows x 131 columns]

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 139273 entries, 0 to 139272
Columns: 131 entries, date to smart_255_raw
dtypes: float64(126), int64(2), object(3)
memory usage: 139.2+ MB


In [12]:
cols = list(df.columns)
cols

['date',
 'serial_number',
 'model',
 'capacity_bytes',
 'failure',
 'smart_1_normalized',
 'smart_1_raw',
 'smart_2_normalized',
 'smart_2_raw',
 'smart_3_normalized',
 'smart_3_raw',
 'smart_4_normalized',
 'smart_4_raw',
 'smart_5_normalized',
 'smart_5_raw',
 'smart_7_normalized',
 'smart_7_raw',
 'smart_8_normalized',
 'smart_8_raw',
 'smart_9_normalized',
 'smart_9_raw',
 'smart_10_normalized',
 'smart_10_raw',
 'smart_11_normalized',
 'smart_11_raw',
 'smart_12_normalized',
 'smart_12_raw',
 'smart_13_normalized',
 'smart_13_raw',
 'smart_15_normalized',
 'smart_15_raw',
 'smart_16_normalized',
 'smart_16_raw',
 'smart_17_normalized',
 'smart_17_raw',
 'smart_18_normalized',
 'smart_18_raw',
 'smart_22_normalized',
 'smart_22_raw',
 'smart_23_normalized',
 'smart_23_raw',
 'smart_24_normalized',
 'smart_24_raw',
 'smart_168_normalized',
 'smart_168_raw',
 'smart_170_normalized',
 'smart_170_raw',
 'smart_173_normalized',
 'smart_173_raw',
 'smart_174_normalized',
 'smart_174_raw

ah, 131 columns, this will be fun.
we have 5 identifying columns and 126 SMART statistics, 63 normalized 63 raw. both could be interesting.

- we have 3 strings:
    - date, serial #, model
- 2 ints:
    - capacity_bytes
    - failure
- the rest floats
    - SMART scores
    
    
Date – The date of the file in yyyy-mm-dd format.

Serial Number – The manufacturer-assigned serial number of the drive.

Model – The manufacturer-assigned model number of the drive.

Capacity – The drive capacity in bytes.

Failure – Contains a “0” if the drive is OK. Contains a “1” if this is the last day the drive was operational before failing.

the good news is that we can just use a scalar encoder.
the bad news is that htm.core doesn't have multiEncoder, so... i'll have to loop through the same encoder and concat into a HUGE SDR, or figure out which features matter more.

trouble is that the max for some of these are in the millions; larger rangers means more bits required to encode said feature. perhaps there are some outliers messing this all up.

In [15]:
df.describe()

capacity_bytes        failure  smart_1_normalized   smart_1_raw  \
count    1.392730e+05  139273.000000       139161.000000  1.391610e+05   
mean     9.105128e+12       0.000022           90.941406  8.714940e+07   
std      3.673838e+12       0.004641           14.610727  8.118958e+07   
min     -1.000000e+00       0.000000           58.000000  0.000000e+00   
25%      4.000787e+12       0.000000           80.000000  0.000000e+00   
50%      1.200014e+13       0.000000           83.000000  7.317012e+07   
75%      1.200014e+13       0.000000          100.000000  1.585940e+08   
max      1.600090e+13       1.000000          200.000000  4.436847e+08   

       smart_2_normalized   smart_2_raw  smart_3_normalized    smart_3_raw  \
count        38813.000000  38813.000000       138553.000000  138553.000000   
mean           121.636101     60.017803           98.007290     484.466854   
std             21.663922     49.837689           15.399402    1743.075542   
min             86.000000      0.000000           75.000000       0.000000   
25%            100.000000      0.000000           91.000000       0.000000   
50%            132.000000     96.000000           95.000000       0.000000   
75%            134.000000    102.000000          100.000000       0.000000   
max            252.000000   1061.000000          253.000000   11070.000000   

       smart_4_normalized    smart_4_raw  ...  smart_250_normalized  \
count       138553.000000  138553.000000  ...                   0.0   
mean            99.998297      10.039422  ...                   NaN   
std              0.114662     123.978045  ...                   NaN   
min             75.000000       1.000000  ...                   NaN   
25%            100.000000       3.000000  ...                   NaN   
50%            100.000000       6.000000  ...                   NaN   
75%            100.000000      10.000000  ...                   NaN   
max            100.000000   25141.000000  ...                   NaN   

       smart_250_raw  smart_251_normalized  smart_251_raw  \
count            0.0                   0.0            0.0   
mean             NaN                   NaN            NaN   
std              NaN                   NaN            NaN   
min              NaN                   NaN            NaN   
25%              NaN                   NaN            NaN   
50%              NaN                   NaN            NaN   
75%              NaN                   NaN            NaN   
max              NaN                   NaN            NaN   

       smart_252_normalized  smart_252_raw  smart_254_normalized  \
count                   0.0            0.0            286.000000   
mean                    NaN            NaN            103.496503   
std                     NaN            NaN             18.401325   
min                     NaN            NaN            100.000000   
25%                     NaN            NaN            100.000000   
50%                     NaN            NaN            100.000000   
75%                     NaN            NaN            100.000000   
max                     NaN            NaN            200.000000   

       smart_254_raw  smart_255_normalized  smart_255_raw  
count          286.0                   0.0            0.0  
mean             0.0                   NaN            NaN  
std              0.0                   NaN            NaN  
min              0.0                   NaN            NaN  
25%              0.0                   NaN            NaN  
50%              0.0                   NaN            NaN  
75%              0.0                   NaN            NaN  
max              0.0                   NaN            NaN  

[8 rows x 128 columns]

### more info from the website regarding interpretation:
Blank Fields

The daily snapshots record the SMART stats information reported by the drive. Since most drives do not report values for all SMART stats, there are blank fields in every record. Also, different drives may report different stats based on their model and/or manufacturer.


Inconsistent Fields

Reported stats for the same SMART stat can vary in meaning based on the drive manufacturer and the drive model. Make sure you are comparing apples-to-apples as drive manufacturers don't generally disclose what their specific numbers mean.


Out-of-Bounds Values

The values in the files are the values reported by the drives. Sometimes, those values are out of whack. For example, in a few cases the RAW value of SMART 9 (Drive life in hours) reported a value that would make a drive 10+ years old, which was not possible. In other words, it’s a good idea to have bounds checks when you process the data.


The Number of Drives Will Change

When a drive fails, the "Failure" field is set to "1" on the day it fails. The next day, the drive is removed from the list and is no longer counted, reducing the overall number of drives. On the other hand, new drives are added on a regular basis increasing the overall number of drives. In other words, count the number of drives each day.


Computing Drive Days

Each day a drive is listed in a daily snapshot file it counts as one drive day. For example, if there are 35,000 drives listed in a daily snapshot file that equals 35,000 drive days. In the docs.zip file you can download below, you’ll find a PDF file named “computing_failure_rates.pdf” which describes how we compute drive days, drive years, and drive failures rates.


Drive Age

As noted, the RAW value of SMART 9 is the number of hours a drive has been in service up to that point. To determine the drive’s age in days, you divide the reported number by 24.


Missing Data

In Q1 2017, the SMART stats for some of the hard drives during the period of January 28 thru 31, 2017 were not recorded. On February 1, 2017 complete reporting resumed. While this had no effect on the how we use the data, it may, depending on how you use the data, effect your efforts.

## important points:
- "apples to apples" SMART stats mean different things from different manufacturers, who don't tell you what the numbers mean. did i read that right? that sounds wacky
    - this should be alright if each manufacturer/model has enough data to "learn" from.
        - however, the "memory_length" of HTM must be questioned here. or perhaps not, since we're not using TM for time patterns.
- out of bounds values: some outliers should be removed
- drive age... SMART 9 is #hours a drive has been in service. should we give this a few more bits than the other vars?
- encoding size
- hotgym used 700 encoding into 2048 SP. but that's determinable based on runtime & feature size tradeoffs
    - determining proper feature encoding size:
        - 1) trim outliers
        - 2) make a list of the numerical range for each variable
            - capacity_bytes could be a category-string encoder, since it's almost unique enough to be... no, probably not. plenty of hard drives of different manufacturers/models will have arbitrary 'round' cutoffs like "1tb" etc.
        - 3) give each of the SMART stats "too many" bits... like 20 or 30. 
        - 4) scale the SDR down to sanity. 
        
- i mean, the first experiment is going to be configuring one or two RDSEs and looping every feature through it + adding that feat_encoding to the growing SDR for each row
    - this'll... probably be bad, but that's good
- the question is: if too few bits are allocated, will it crash/let me know, or will different data be encoded as similar (breakdown of encoder principles - ML dyslexia, of a sort)?

In [17]:
df.sort_values(by=['serial_number']).head(200)

date     serial_number                model  capacity_bytes  \
127130  2020-05-27  0564f6f3fab90010          DELLBOSS VD    480036847616   
85950   2020-05-27      1050A03NF97G  TOSHIBA MG07ACA14TA  14000519643136   
123631  2020-05-27      1050A04MF97G  TOSHIBA MG07ACA14TA  14000519643136   
8548    2020-05-27      1080A0AMF97G  TOSHIBA MG07ACA14TA  14000519643136   
71073   2020-05-27      1080A0ARF97G  TOSHIBA MG07ACA14TA  14000519643136   
...            ...               ...                  ...             ...   
49180   2020-05-27      10A0A08PF97G  TOSHIBA MG07ACA14TA  14000519643136   
11111   2020-05-27      10A0A08QF97G  TOSHIBA MG07ACA14TA  14000519643136   
53725   2020-05-27      10A0A08RF97G  TOSHIBA MG07ACA14TA  14000519643136   
46229   2020-05-27      10A0A08SF97G  TOSHIBA MG07ACA14TA  14000519643136   
27159   2020-05-27      10A0A08TF97G  TOSHIBA MG07ACA14TA  14000519643136   

        failure  smart_1_normalized  smart_1_raw  smart_2_normalized  \
127130        0                 NaN          NaN                 NaN   
85950         0               100.0          0.0               100.0   
123631        0               100.0          0.0               100.0   
8548          0               100.0          0.0               100.0   
71073         0               100.0          0.0               100.0   
...         ...                 ...          ...                 ...   
49180         0               100.0          0.0               100.0   
11111         0               100.0          0.0               100.0   
53725         0               100.0          0.0               100.0   
46229         0               100.0          0.0               100.0   
27159         0               100.0          0.0               100.0   

        smart_2_raw  smart_3_normalized  ...  smart_250_normalized  \
127130          NaN                 NaN  ...                   NaN   
85950           0.0               100.0  ...                   NaN   
123631          0.0               100.0  ...                   NaN   
8548            0.0               100.0  ...                   NaN   
71073           0.0               100.0  ...                   NaN   
...             ...                 ...  ...                   ...   
49180           0.0               100.0  ...                   NaN   
11111           0.0               100.0  ...                   NaN   
53725           0.0               100.0  ...                   NaN   
46229           0.0               100.0  ...                   NaN   
27159           0.0               100.0  ...                   NaN   

        smart_250_raw  smart_251_normalized  smart_251_raw  \
127130            NaN                   NaN            NaN   
85950             NaN                   NaN            NaN   
123631            NaN                   NaN            NaN   
8548              NaN                   NaN            NaN   
71073             NaN                   NaN            NaN   
...               ...                   ...            ...   
49180             NaN                   NaN            NaN   
11111             NaN                   NaN            NaN   
53725             NaN                   NaN            NaN   
46229             NaN                   NaN            NaN   
27159             NaN                   NaN            NaN   

        smart_252_normalized  smart_252_raw  smart_254_normalized  \
127130                   NaN            NaN                   NaN   
85950                    NaN            NaN                   NaN   
123631                   NaN            NaN                   NaN   
8548                     NaN            NaN                   NaN   
71073                    NaN            NaN                   NaN   
...                      ...            ...                   ...   
49180                    NaN            NaN                   NaN   
11111                    NaN            NaN                   NaN   
53725            

In [24]:
df['serial_number'].nunique()

139273

In [25]:
df['model'].nunique()

45

12673843

# dropping null columns
a lot of columns have mostly nulls. but some have ~10% null - should we drop those or keep them in the encoding?

## we should also take care not to drop any columns with failure=1, since we've got such a class imbalance to deal with

- this guy (https://datatonic.com/insights/machine-learning-predictive-maintenance-hard-drive-failure/) deals with the class imbalance by adding in a 'failure horizon': 
        - As a first step to address these problems we can define different “failure horizons” of N days before the labelled failure such that all instances during these days are also labelled as failures. If we define a failure horizon of a week, for example, then the 7 days before the actual failure will also be labelled failures
- via the logic that if a drive fails on November 14th, it was probably not doing so well - already on the road to failure, one might say - since November 7th, so fill in the last 6 or 7 days with "1" as well.
    - interesting solution. SMART stats are just readings - they don't make any claims about the imminent failure of the PC (that's our job). 
    - this is actually quite clever, i like this a lot. i shall steal the idea, of course.
- he also uses the entire dataset and zeroes in on one specific model of hard drive -> not production-ready, wants to test validity
- for continuous cortical algos, i think we don't have to constrain our initial run to single models, really. that's the advantage of all this - huge runtime and energy gains compared to classical deep learning
    - but he tests different 'failure horizons' of various day-lengths (rows, since daily measurements), which is quite savvy. if you're testing an experiment that drastically alters the data & model results, test it in several different directions.



does htm.core crash with NaNs? if so i'll just zero the columns

In [33]:
df.isna().sum()

date                         0
serial_number                0
model                        0
capacity_bytes               0
failure                      0
                         ...  
smart_252_raw           139273
smart_254_normalized    138987
smart_254_raw           138987
smart_255_normalized    139273
smart_255_raw           139273
Length: 131, dtype: int64

In [37]:
nulls = list(df.isna().sum())

In [39]:
# align with cols
nul = dict(zip(cols, nulls))
nul

{'date': 0,
 'serial_number': 0,
 'model': 0,
 'capacity_bytes': 0,
 'failure': 0,
 'smart_1_normalized': 112,
 'smart_1_raw': 112,
 'smart_2_normalized': 100460,
 'smart_2_raw': 100460,
 'smart_3_normalized': 720,
 'smart_3_raw': 720,
 'smart_4_normalized': 720,
 'smart_4_raw': 720,
 'smart_5_normalized': 720,
 'smart_5_raw': 720,
 'smart_7_normalized': 720,
 'smart_7_raw': 720,
 'smart_8_normalized': 100460,
 'smart_8_raw': 100460,
 'smart_9_normalized': 112,
 'smart_9_raw': 112,
 'smart_10_normalized': 720,
 'smart_10_raw': 720,
 'smart_11_normalized': 138539,
 'smart_11_raw': 138539,
 'smart_12_normalized': 112,
 'smart_12_raw': 112,
 'smart_13_normalized': 139273,
 'smart_13_raw': 139273,
 'smart_15_normalized': 139273,
 'smart_15_raw': 139273,
 'smart_16_normalized': 138665,
 'smart_16_raw': 138665,
 'smart_17_normalized': 138665,
 'smart_17_raw': 138665,
 'smart_18_normalized': 122038,
 'smart_18_raw': 122038,
 'smart_22_normalized': 124803,
 'smart_22_raw': 124803,
 'smart_23_n

## 98/126 SMART cols have >20% nulls

In [47]:
twent = len(df)/5
print(twent)
c = 0
for col, null in nul.items():
    if null > twent:
        print(null)
        c+=1
c

27854.6
100460
100460
100460
100460
138539
138539
139273
139273
139273
139273
138665
138665
138665
138665
122038
122038
124803
124803
132053
132053
132053
132053
138665
138665
138665
138665
138665
138665
138665
138665
138665
138665
139273
139273
139273
139273
139273
139273
119251
119251
93488
93488
39799
39799
39799
39799
93488
93488
39799
39799
84771
84771
61200
61200
100194
100194
83650
83650
139273
139273
138665
138665
131130
131130
131130
131130
128062
128062
131130
131130
138805
138805
131130
131130
138665
138665
138665
138665
138665
138665
138665
138665
31585
31585
36581
36581
36581
36581
139273
139273
139273
139273
139273
139273
138987
138987
139273
139273


98

In [ ]:
fif = len(df)/10
c = 0
for col, null in nul.items():
    if null > twent:
        print(null)
        c+=1
c